### PR TITLE
Refactor ads::ResultCallback to base::OnceCallback.

### PR DESCRIPTION
**Pull request by aseren**
Resolves https://github.com/brave/brave-browser/issues/24761

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/ads/ad_events/search_result_ads/search_result_ad_event_handler.cc [hunk 1]
```diff
@@ -59,10 +62,11 @@ void EventHandler::RemoveObserver(EventHandlerObserver* observer) {
   observers_.RemoveObserver(observer);
 }
 
-void EventHandler::FireEvent(const mojom::SearchResultAdInfoPtr& ad_mojom,
+void EventHandler::FireEvent(mojom::SearchResultAdInfoPtr ad_mojom,
                              const mojom::SearchResultAdEventType event_type,
                              FireAdEventHandlerCallback callback) const {
   DCHECK(mojom::IsKnownEnumValue(event_type));
+  DCHECK(ad_mojom);
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/ad_events/search_result_ads/search_result_ad_event_handler.cc:69]: nit: same order as params

File: vendor/bat-native-ads/src/bat/ads/internal/ads/ad_events/search_result_ads/search_result_ad_event_handler_unittest.cc [hunk 2]
```diff
@@ -114,10 +115,10 @@ class BatAdsSearchResultAdEventHandlerTest : public EventHandlerObserver,
     UnitTestBase::TearDown();
   }
 
-  void FireEvent(const mojom::SearchResultAdInfoPtr& ad_mojom,
+  void FireEvent(mojom::SearchResultAdInfoPtr ad_mojom,
                  const mojom::SearchResultAdEventType event_type) {
     event_handler_->FireEvent(
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/ad_events/search_result_ads/search_result_ad_event_handler_unittest.cc:120]: DCHECK for both ad_mojom and event_type

File: vendor/bat-native-ads/src/bat/ads/internal/ads/ad_events/search_result_ads/search_result_ad_event_handler.cc [hunk 3]
```diff
@@ -110,67 +114,81 @@ void EventHandler::FireEvent(const SearchResultAdInfo& ad,
...
   const DepositInfo deposit = BuildDeposit(ad_mojom);
 
+  database::table::Deposits deposits_database_table;
+  deposits_database_table.Save(
+      deposit,
+      base::BindOnce(&EventHandler::OnSaveDeposits, base::Unretained(this),
+                     std::move(ad_mojom), callback));
+}
+
+void EventHandler::OnSaveDeposits(mojom::SearchResultAdInfoPtr ad_mojom,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/ad_events/search_result_ads/search_result_ad_event_handler.cc:130]: DCHECK ad_mojom

**Review by tmancey — APPROVED** [review]
LGTM++

File: vendor/bat-native-ads/src/bat/ads/internal/ads/ad_events/search_result_ads/search_result_ad_event_handler_unittest.cc (hunk 2 above)
Comment (aseren) [reply vendor/bat-native-ads/src/bat/ads/internal/ads/ad_events/search_result_ads/search_result_ad_event_handler_unittest.cc:120]: I think these DCHECKs are unnecessary here, as we are just passing these parameters to EventHandler::FireEvent, and there is already DCHECK there.

**Pull request merged**